cg/store/crud/read.py

```diff
@@ -1604,3 +1637,50 @@ def get_related_cases(
                 CaseFilter.BY_CUSTOMER_ENTRY_IDS,
             ],
         ).all()
+
+    def get_uploaded_related_dna_cases(self, rna_case: Case) -> list[Case]:
+        """Returns all uploaded DNA cases related to the specified RNA case."""
+        dna_cases: set[Case] = self.get_related_dna_cases_from_rna_case(rna_case=rna_case)
+        uploaded_dna_cases: list[Case] = []
+        for dna_case in dna_cases:
+            if dna_case.is_uploaded:
+                uploaded_dna_cases.append(dna_case)
+            else:
+                LOG.warning(f"Related DNA case {dna_case.internal_id} has not been completed.")
+        return uploaded_dna_cases
+
+    def get_related_dna_cases_from_rna_case(self, rna_case: Case) -> set[Case]:
+        """Return a list of unique DNA cases related to the samples of an RNA case."""
+        dna_cases: set[Case] = set()
+        for sample in rna_case.samples:
+            dna_cases.update(self._get_related_dna_cases_from_rna_sample(sample))
+        return dna_cases
+
+    def _get_related_dna_cases_from_rna_sample(self, rna_sample: Sample) -> list[Case]:
+        """Return a list of DNA cases related to an RNA sample."""
+
+        if not rna_sample.subject_id:
+            raise CgDataError(
+                f"Failed to link RNA sample {rna_sample.internal_id} to DNA samples - subject_id field is empty."
+            )
+
+        collaborators: set[Customer] = rna_sample.customer.collaborators
+        related_dna_samples: list[Sample] = self.get_related_samples(
+            sample_internal_id=rna_sample.internal_id,
+            prep_categories=DNA_PREP_CATEGORIES,
+            collaborators=collaborators,
+        )
+
+        if len(related_dna_samples) != 1:
+            raise CgDataError(
+                f"Unexpected number of DNA sample matches for subject_id: "
+                f"{rna_sample.subject_id}. Number of matches: {len(related_dna_samples)} "
+            )
+        dna_sample: Sample = related_dna_samples[0]
+
+        dna_cases: list[Case] = self.get_related_cases(
+            sample_internal_id=dna_sample.internal_id,
+            workflows=DNA_WORKFLOWS_WITH_SCOUT_UPLOAD,
+            collaborators=collaborators,
+        )
+        return dna_cases
```

Comment:
```suggestion
    def get_uploaded_dna_cases_related_to_rna_case(self, rna_case_id: str, prep_categories: list[str], workflows: list[str]) -> list[Case]:
    """Get all uploaded DNA cases related to a given RNA case using a database query."""
    return (
        self.session.query(Case)
        .join(Sample, Sample.case_id == Case.id)
        .filter(
            Sample.prep_category.in_(prep_categories),
            Sample.subject_id.isnot(None),
            Case.workflow.in_(workflows),
            Case.is_uploaded.is_(True),
            Case.rna_case_id == rna_case_id
        )
        .all()
    )
```

Reply:
We do not need to loop though cases or samples, we likely can get the answer that we want from the database with a direct query. 

Copying the same functions from the service into the crud does not solve the efficiency problem with extracting the data and looping of the database rows in python.